client/modules/IDE/reducers/files.js

```diff
@@ -159,8 +160,22 @@ const files = (state, action) => {
         return Object.assign({}, file, { blobURL: action.blobURL });
       });
     case ActionTypes.NEW_PROJECT:
+      action.files = action.files.map((file) => {
+        const corrospondingObj = state.find((obj) => obj.id === file.id);
+        if (corrospondingObj && corrospondingObj.fileType === 'folder') {
+          file.isFolderClosed = corrospondingObj.isFolderClosed;
+        }
+        return file;
+      });
       return setFilePaths(action.files);
     case ActionTypes.SET_PROJECT:
+      action.files = action.files.map((file) => {
+        const corrospondingObj = state.find((obj) => obj.id === file.id);
+        if (corrospondingObj && corrospondingObj.fileType === 'folder') {
+          file.isFolderClosed = corrospondingObj.isFolderClosed;
```

Comment:
You can avoid mutating the action like this:
```
case ActionTypes.SET_PROJECT:
    const files = action.files.map((file) => {
        const corrospondingObj = state.find((obj) => obj.id === file.id);
        if (corrospondingObj && corrospondingObj.fileType === 'folder') {
            const isFolderClosed = corrospondingObj.isFolderClosed || false;
            return { ...file, isFolderClosed };
        }
        return file;
    });
    return setFilePaths(files);
```

FYI there are some mutations in the existing code in this reducer😢.  But let's not add any new ones.

I have to look at the code to remind myself what the difference is between `NEW_PROJECT` and `SET_PROJECT`.  If we need to use the same code in both places then we can move it to a helper function.  But I suspect that it's not needed for the `NEW_PROJECT` case because if it's a new project then the `corrospondingObj` would always be `undefined`.

Reply:
>I suspect that it's not needed for the NEW_PROJECT case because if it's a new project then the corrospondingObj would always be undefined.

I was wrong.  It is needed in both places.  But it seems that `NEW_PROJECT` and `SET_PROJECT` are identical so now I'm wondering why we have two actions that do the same thing.

Reply:
Hey @lindapaiste , Thank you for your invaluable and clear response . I have made the suitable changes.